projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php

```diff
@@ -107,15 +107,29 @@ public function prepare_menu_for_response( array $menu ) {
 		 * See: https://github.com/Automattic/wp-calypso/blob/ebde236ec9b21ea9621c0b0523bd5ea185523731/client/state/admin-menu/schema.js
 		 */
 		foreach ( $menu as $menu_item ) {
-			$item = $this->prepare_menu_item( $menu_item );
+
+			/**
+			 * Logic extracted from _wp_menu_output() core function
+			 * Based $admin_is_parent we determine if the submenu items should be pointing to the admin.php
+			 */
+			$menu_hook = get_plugin_page_hook( $menu_item[2], 'admin.php' );
+			$menu_file = $menu_item[2];
+			$pos       = strpos( $menu_file, '?' );
+
+			if ( false !== $pos ) {
+				$menu_file = substr( $menu_file, 0, $pos );
+			}
+
+			$admin_is_parent = ! empty( $menu_hook ) || ( ( 'index.php' !== $menu_item[2] ) && file_exists( WP_PLUGIN_DIR . "/$menu_file" ) && ! file_exists( ABSPATH . "/wp-admin/$menu_file" ) );
```

Comment:
Could this block of code be moved inside of `prepare_menu_item_url()`? It looks like `prepare_menu_item()` and `prepare_submenu_item()` are just passing it through anyway?

Reply:
I think it might add additional complexity for menus that don't have submenus. 

In that case, the[ the $parent_slug is an empty string](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php#L250) and we'll need to pass an additional parameter anyway or add some kind of additional logic to handle that scenario (but I don't think it's a good idea because there might be a risk diverging again from how the menus are built in Core).

Reply:
In the case of an empty `$parent_slug`, couldn't `$admin_is_parent` be false? Something like

```php
$admin_is_parent = false;
if ( ! empty( $parent_slug ) ) {
	$menu_hook = get_plugin_page_hook( $parent_slug, 'admin.php' );
	$menu_file = $parent_slug;
	$pos       = strpos( $menu_file, '?' );

	if ( false !== $pos ) {
		$menu_file = substr( $menu_file, 0, $pos );
	}

	$admin_is_parent = ! empty( $menu_hook ) || ( ( 'index.php' !== $parent_slug ) && file_exists( WP_PLUGIN_DIR . "/$menu_file" ) && ! file_exists( ABSPATH . "/wp-admin/$menu_file" ) );
}
```

Reply:
`$parent_slug` is always empty for menus that don't have submenus, but not all of them have `$admin_is_parent` false; while  `$parent_slug` is indeed `$menu_item[2]`, that's not always the case.

Hmm... I guess we could assume that, when the $parent_slug is empty, we are in the context of a menu and use the $url variable as the menu_item, since for menus [without submenus $url is $menu_item[2].](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php#L249) It would look something like this:
```
       $menu_item = '' !== $parent_slug ? $parent_slug : $url;
	$menu_hook = get_plugin_page_hook( $menu_item, 'admin.php' );
	$menu_file = $menu_item;
	$pos       = strpos( $menu_file, '?' );

	if ( false !== $pos ) {
		$menu_file = substr( $menu_file, 0, $pos );
	}

	$admin_is_parent = ! empty( $menu_hook ) || ( ( 'index.php' !== $parent_slug ) && file_exists( WP_PLUGIN_DIR . "/$menu_file" ) && ! file_exists( ABSPATH . "/wp-admin/$menu_file" ) );

```
However, it feels like this is unreliable on the long term and adds more complexity because it introduces a hidden behaviour that's not clear which scenario it handles.

Reply:
I think it's fine to leave this piece of logic in the current place. That way it'll be more aligned with Core, which determines the value of that variable before iterating the submenu items.

Reply:
I'd be disappointed if we saw the menu generation code in Core as an ideal to strive for. As long as the outcome is similar we should be free to improve it. `prepare_menu_item()` and `prepare_submenu_item()` seem to be passing `$admin_is_parent` through untouched. Is there a functional reason I'm missing why it can't live in `prepare_menu_item_url()`?

Reply:
> Is there a functional reason I'm missing why it can't live in `prepare_menu_item_url()`?

After taking a further look at what Core does, I noted that `$admin_is_parent` is only checked for submenu items, so it makes sense to set it to false when we're certain that the given URL in `prepare_menu_item_url()` belongs to a parent menu item.

However, we cannot trivially determine in `prepare_menu_item_url()` whether the given URL is for a parent menu item, because `$parent_slug` is defined for both [submenu items](https://github.com/Automattic/jetpack/blob/991fab98c8989618d895e676f311a54c052b2da0/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php#L299) **AND** [parent menu items](https://github.com/Automattic/jetpack/blob/991fab98c8989618d895e676f311a54c052b2da0/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php#L252-L255) (it is only empty [when the menu doesn't have any submenu item](https://github.com/Automattic/jetpack/blob/991fab98c8989618d895e676f311a54c052b2da0/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php#L252-L255)).

In order to reliably determine if the given URL belongs to a parent menu item, we should add logic that performs some checks in `$parent_slug` based on how that variable is defined in `prepare_menu_item()` and `prepare_submenu_item()`. 

I agree with @BogdanUngureanu above that this approach would be adding unnecessary complexity.

Reply:
> After taking a further look at what Core does, I noted that $admin_is_parent is only checked for submenu items, so it makes sense to set it to false when we're certain that the given URL in prepare_menu_item_url() belongs to a parent menu item.

Said that, I think we could move this logic to `prepare_submenu_item()` since it's the only place where we need it.